grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -51,22 +64,128 @@ class JSONResponse(werkzeug_wrappers_json.JSONMixin,
 
   def __init__(self, response=None, *args, **kwargs) -> None:
     kwargs["mimetype"] = JSON_MIME_TYPE
-    if response is not None and not isinstance(response, werkzeug_wsgi.ClosingIterator):
+    if response is not None and not isinstance(response,
+                                               werkzeug_wsgi.ClosingIterator):
       response = json.dumps(response)
     super().__init__(response, *args, **kwargs)
 
 
+class Metric(ABC):
+  """A single metric that can be fetched from a
+  Fleetspeak-based GRR deployment."""
+
+  @abstractmethod
+  def __init__(self, name: str) -> None:
+    """Initializes a new metric."""
+    self.name = name
+
+  @abstractmethod
+  def ProcessQuery(self, req: JSONRequest):
```

Comment:
Missing return type.

Reply:
That's something I was very divided about. The two implementations of `Metric` (`ClientResourceUsageMetric` and `ClientsStatisticsMetric`) have different return types in the method `ProcessQuery`.
I tried creating a parent class for the named tuples `_TargetWithDatapoints` and `_TableQueryResult`, but [results I found online](https://github.com/python/typing/issues/427) do not suggest there is a good way of doing this (besides using dataclasses).

What do you think?

Reply:
Can you use [Union](https://docs.python.org/3/library/typing.html#typing.Union)?

Reply:
> Can you use Union?

I would refrain from using `Union`, if possible. They can be useful for easy typifying of existing code, but usually they indicate poor design.

Here, if you look at how your abstract method is used it is easy to understand how to type it:

```python
targets_with_datapoints = [
    AVAILABLE_METRICS_DICT[target].ProcessQuery(json_data)
    for target in requested_targets
]
response = [t._asdict() for t in targets_with_datapoints]
return JSONResponse(response=response)
```

You don't care about a particular type, you care that you can call `_asdict()` on the results. Hence, it is enough to say that `ProcessQuery` should return a `NamedTuple` instance. Whether it is `ClientResourceUsageMetric` or `ClientsStatisticsMetric` doesn't matter.

Reply:
That makes sense, the problem I get when I try to specify `NamedTuple` as a return type for `ProcessQuery`, mypy generates the following errors:
```
grrafana.py:112: error: Return type "_TargetWithDatapoints" of "ProcessQuery" incompatible with return type "NamedTuple" in supertype "Metric"
grrafana.py:136: error: Return type "_TableQueryResult" of "ProcessQuery" incompatible with return type "NamedTuple" in supertype "Metric"
```
It's weird as both `_TargetWithDatapoints` and `_TableQueryResult` inherit from `NamedTuple`.

Reply:
Hmm, you are right. Looks like this is a known issue: python/mypy#3915, python/typing#431. The code seems to work under pytype, though. I would still try to annotate it with `NamedTuple`, add `# type: ignore[return-value]` to suppress the warnings and add a `TODO` comment with links to the mentioned tracking issues.

Reply:
> Hmm, you are right. Looks like this is a known issue: [python/mypy#3915](https://github.com/python/mypy/issues/3915), [python/typing#431](https://github.com/python/typing/issues/431). The code seems to work under pytype, though. I would still try to annotate it with `NamedTuple`, add `# type: ignore[return-value]` to suppress the warnings and add a `TODO` comment with links to the mentioned tracking issues.

Will do. Just to clarify; the error is actually an [override](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-validity-of-overrides-override) code and not a return-value code.